docs/tutorials/data_provider.md

```diff
@@ -61,3 +61,170 @@ impl AdditiveIdentity {
     }
 }
 ```
+
+## Caching Data Provider
+
+ICU4X has no internal caches because there is no one-size-fits-all solution. It is easy for clients to implement their own cache for ICU4X, and although this is not generally required or recommended, it may be beneficial when latency is of utmost importance and, for example, a less-efficient data provider such as JSON is being used.
+
+The following example illustrates an LRU cache on top of a BufferProvider that saves deserialized data payloads as type-erased objects and then checks for a cache hit before calling the inner provider.
+
+```rust
+use icu_provider::hello_world::HelloWorldFormatter;
+use icu_provider::prelude::*;
+use icu::locid::locale;
+use lru::LruCache;
+use std::borrow::{Borrow, Cow};
+use std::convert::TryInto;
+use std::sync::Mutex;
+use yoke::trait_hack::YokeTraitHack;
+use yoke::Yokeable;
+use zerofrom::ZeroFrom;
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct CacheKeyWrap(CacheKey<'static>);
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct CacheKey<'a>(DataKey, Cow<'a, DataLocale>);
+
+pub struct LruDataCache<P> {
+    cache: Mutex<LruCache<CacheKeyWrap, AnyResponse>>,
+    provider: P,
+}
+
+// This impl enables a borrowed DataLocale to be used during cache retrieval.
+impl<'a> Borrow<CacheKey<'a>> for lru::KeyRef<CacheKeyWrap> {
+    fn borrow(&self) -> &CacheKey<'a> {
+        &Borrow::<CacheKeyWrap>::borrow(self).0
+    }
+}
+
+impl<M, P> DataProvider<M> for LruDataCache<P>
+where
+    M: KeyedDataMarker + 'static,
+    M::Yokeable: ZeroFrom<'static, M::Yokeable>,
+    M::Yokeable: icu_provider::MaybeSendSync,
+    for<'a> YokeTraitHack<<M::Yokeable as Yokeable<'a>>::Output>: Clone,
+    P: DataProvider<M>,
+{
+    fn load(&self, req: DataRequest) -> Result<DataResponse<M>, DataError> {
+        {
+            // First lock: cache retrieval
+            let mut cache = self.cache.lock().unwrap();
+            let borrowed_cache_key = CacheKey(M::KEY, Cow::Borrowed(req.locale));
+            if let Some(any_res) = cache.get(&borrowed_cache_key) {
+                // Note: Cloning a DataPayload is usually cheap, and it is necessary in order to
+                // convert the short-lived cache object into one we can return.
+                return any_res.downcast_cloned();
+            }
+        }
+        // Release the lock to invoke the inner provider
+        let response: DataResponse<M> = self.provider.load(req)?;
```

Comment:
Is this type needed? `self.provider` is only `DataProvider<M>`, so there shouldn't be any ambiguity.

Reply:
Primarily for clarity. The audience for this tutorial doesn't know what all these functions do or what they return. But if you think it's clear enough I'll remove it.